src/mina-signer/signature.ts

```diff
@@ -0,0 +1,207 @@
+ * which `sign` re-derives by scaling the Pallas "Mina" generator by the `privateKey`.
+ */
+function deriveNonce(
+  message: HashInput,
+  publicKey: Group,
+  privateKey: Scalar,
+  networkId: NetworkId
+): Scalar {
+  let { x, y } = publicKey;
+  let d = Field.fromBits(Scalar.toBits(privateKey));
+  let id = networkId === 'mainnet' ? networkIdMainnet : networkIdTestnet;
+  let input = HashInput.append(message, {
+    fields: [x, y, d],
+    packed: [[id, 8]],
+  });
+  let packedInput = packToFields(input);
+  let inputBits = packedInput.map(Field.toBits).flat();
+  let inputBytes = bitsToBytes(inputBits);
+  let bytes = blake2b(Uint8Array.from(inputBytes), undefined, 32);
+  // drop the top two bits to convert into a scalar field element
+  // (creates negligible bias because q = 2^254 + eps, eps << q)
+  bytes[bytes.length - 1] &= 0x3f;
+  return Scalar.fromBytes([...bytes]);
+}
+
+/**
+ * Hash a message for use by the Schnorr signature algorithm, by:
+ * - packing the inputs `message`, `publicKey`, `r` into an array of Pallas base field elements,
+ * - apply a salted hash with the {@link Poseidon} hash function,
+ * - interpreting the resulting base field element as a scalar
+ *   (which is always possible, and is a no-op, since the scalar field is larger and both fields are represented with bigints).
+ *
+ * @see {@link https://github.com/MinaProtocol/mina/blob/develop/docs/specs/signatures/description.md detailed spec of the algorithm}
+ *
+ * In contrast to the spec above, this uses the "chunked" style of hash input packing, implemented in {@link packToFields}.
+ *
+ * @param message an arbitrary {@link HashInput}
+ * @param publicKey an affine, non-zero point on the Pallas curve, derived by {@link sign} from the private key
+ * @param r an element of the Pallas base field, computed by {@link sign} as the x-coordinate of the generator, scaled by the nonce.
+ * @param networkId either "testnet" or "mainnet", determines the salt (initial state) in the Poseidon hash.
+ */
+function hashMessage(
+  message: HashInput,
+  publicKey: Group,
+  r: Field,
+  networkId: NetworkId
+): Scalar {
+  let { x, y } = publicKey;
+  let input = HashInput.append(message, { fields: [x, y, r] });
+  let prefix =
+    networkId === 'mainnet'
+      ? prefixes.signatureMainnet
+      : prefixes.signatureTestnet;
+  return hashWithPrefix(prefix, packToFields(input));
+}
+
+/**
+ * Verifies a signature created by {@link sign}, returns `true` if (and only if) the signature is valid.
+ *
+ * @see {@link https://github.com/MinaProtocol/mina/blob/develop/docs/specs/signatures/description.md detailed spec of the algorithm}
+ *
+ * In contrast to the spec above, this uses the "chunked" style of hash input packing, implemented in {@link packToFields}.
+ *
+ * @param publicKey the public key has to be passed in as a compressed {@link PublicKey}.
+ * It can be created from a base58 string with {@link PublicKey.fromBase58}.
+ */
+function verify(
```

Comment:
Baring question about infinity check below, this implementation looks correct to me.